base_layer/core/src/block.rs

```diff
@@ -51,29 +65,69 @@ impl Block {
     /// This function will check the block to ensure that all UTXO's are validly constructed and that all signatures are
     /// valid. It does _not_ check that the inputs exist in the current UTXO set;
     /// nor does it check that the PoW is the largest accumulated PoW value.
-    pub fn check_internal_consistency(&self) -> Result<(), TransactionError> {
-        let mut trans: Transaction = self.body.clone().into(); // todo revisit this one q=whole code chain is completed
+    pub fn check_internal_consistency(&self) -> Result<(), BlockError> {
+        self.check_kernel_rules()?;
+        self.check_spent_utxo_rules()?;
+        let mut trans: Transaction = self.body.clone().into();
         trans.offset = self.header.total_kernel_offset.clone();
+        // We need to add the coinbase off set to the block otherwise it will not balance.
+        self.add_coinbase_offset(&mut trans);
+        // no lets check the block
         trans.validate_internal_consistency(&PROVER, &COMMITMENT_FACTORY)?;
         self.check_pow()
     }
 
-    pub fn check_pow(&self) -> Result<(), TransactionError> {
+    /// This function will check spent utxo rules like coinbase maturity height etc
+    pub fn check_spent_utxo_rules(&self) -> Result<(), BlockError> {
+        // for txo in &self.body.inputs{
+        //     if txo.features.contains(OutputFeatures::COINBASE_OUPUT){
+        // ToDo fill this out in a seperate PR.  as this requires extra code and modifications to the MMR,
+        // we currently cannot get the creation height of a given  commitment.
+        //     }
+        // }
         Ok(())
     }
 
+    /// This function will check spent kernel rules like tx lock height etc
+    pub fn check_kernel_rules(&self) -> Result<(), BlockError> {
+        for kernel in &self.body.kernels {
+            if kernel.lock_height > self.header.height {
+                return Err(BlockError::InvalidKernel);
+            }
+        }
+        Ok(())
+    }
+
+    /// This function will check that the supplied PoW is valid
+    pub fn check_pow(&self) -> Result<(), BlockError> {
+        self.header
+            .pow
+            .validate_pow(&self.body)
+            .map_err(BlockError::ProofOfWorkError)
+    }
+
     /// This function will calculate the pow for the block and fill out the pow header field
     pub fn calculate_pow(&mut self) -> Result<(), TransactionError> {
         // todo
         Ok(())
     }
+
+    // Add an offset for the coinbase to the transaction
+    fn add_coinbase_offset(&self, trans: &mut Transaction) {
```

Comment:
The `add` in the method name sounds like it alters the state of `self`. It doesn't actually look like it needs much from the Block other than height. Perhaps it's better to put this in Transaction?

Reply:
This relies on the `calculate_coinbase` function. I dont want to make transaction dependant on that. Transactions doesnt need to know about the emission schedule

Reply:
How about creating a `ConsensusRules` struct that handles calculating coinbase and then having something like this:
`trans.add_coinbase_offset(rules.calculate_emission(...))?;` ? 

I was surprised that a method with a mutating verb i.e. "add" didn't take a `&mut self`, and since the `trans` is  `&mut` it seemed like it might be better placed on the Transaction

Reply:
Perhaps a change in name of the function? I agree that trans.add_coinbase_offset is more clear what happens then block.add_coinbase_offset. But transaction still should not care about emissions or stuff like that. 

The only reason I am doing this this way is to leverage the already existing code to check a balancing transaction and in essence a block is a transaction + emission.